content_scripts/marks.coffee

```diff
@@ -1,45 +1,77 @@
-      localStorage["vimiumMark|#{baseLocation}|#{keyChar}"] = JSON.stringify
-        scrollX: window.scrollX,
-        scrollY: window.scrollY
-      HUD.showForDuration "Created local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
-
-root.activateGotoMode = ->
-  handlerStack.push keydown: (e) ->
-    keyChar = KeyboardUtils.getKeyChar(event)
-    return unless keyChar isnt ""
-
-    if /[A-Z]/.test keyChar
-      chrome.runtime.sendMessage
-        handler: 'gotoMark'
-        markName: keyChar
-    else if /[a-z]/.test keyChar
-      [baseLocation, sep, hash] = window.location.href.split '#'
-      markString = localStorage["vimiumMark|#{baseLocation}|#{keyChar}"]
-      if markString?
-        mark = JSON.parse markString
-        window.scrollTo mark.scrollX, mark.scrollY
-        HUD.showForDuration "Jumped to local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
+
+Marks =
+  mode: null
+  previousPosition: null
+
+  exit: (continuation = null) ->
+    @mode?.exit()
+    @mode = null
+    continuation?()
+
+  # This returns the key which is used for storing mark locations in localStorage.
+  getLocationKey: (keyChar) ->
+    "vimiumMark|#{window.location.href.split('#')[0]}|#{keyChar}"
+
+  getMarkString: ->
+    JSON.stringify scrollX: window.scrollX, scrollY: window.scrollY
+
+  setPreviousPosition: ->
+    @previousPosition = @getMarkString()
+
+  showMessage: (message, keyChar) ->
+    HUD.showForDuration "#{message} \"#{keyChar}\".", 1000
+
+  activateCreateMode: ->
+    @mode = new Mode
+      name: "create-mark"
+      indicator: "Create mark..."
+      exitOnEscape: true
+      suppressAllKeyboardEvents: true
+      keypress: (event) =>
+        keyChar = String.fromCharCode event.charCode
+        # If <Shift> is depressed, then it's a global mark, otherwise it's a local mark.  This is consistent
+        # vim's [A-Z] for global marks, [a-z] for local marks.  However, it also admits other non-Latin
+        # characters.
+        @exit =>
+          if event.shiftKey
+              chrome.runtime.sendMessage
+                handler: 'createMark'
+                markName: keyChar
+                scrollX: window.scrollX
+                scrollY: window.scrollY
+              , => @showMessage "Created global mark", keyChar
+          else
+              localStorage[@getLocationKey keyChar] = @getMarkString()
+              @showMessage "Created local mark", keyChar
+
+  activateGotoMode: (registryEntry) ->
+    # We pick off the last character of the key sequence used to launch this command. Usually this is just "`".
+    # We then use that character, so together usually the sequence "``", to jump back to the previous
+    # position.  The "previous position" is recorded below, and is registered via @setPreviousPosition()
+    # elsewhere for various other jump-like commands.
+    previousPositionKey = registryEntry.key[registryEntry.key.length-1..]
```

Comment:
Vimium is not vim, so we should use vim-like concepts where appropriate, but do what's right in the browser at the end of the day.  I'm not sure the distinction you're making is particularly important for us.

I'll accept that this is a bit of a cunning trick.  I wouldn't even consider it if it wasn't easily discoverable for many users (which it is, because with the default bindings it just works, and does pretty much what any vim user would expect it to do).

In fact, it was only when I thought of this trick that I entertained the idea of doing this at all.  This is not important enough for a new option (or even a new command, for that matter).  However, imagine for a moment that we did add a new command.  Because <tt>\`</tt> is already a binding, it would be impossible to map the new command to <tt>\`\`</tt>.  That's why I think this is the way to go.

So, for most users, it just works as they'd expect.  And we can add a wiki entry for those curious enough.